File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-doc-5-4/ProductComponentIdentifier.cpp
// ProductComponentIdentifier.cpp 
#include "ProductComponentIdentifier.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

ProductComponentIdentifier::ProductComponentIdentifier(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //premiumProductReferenceNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* premiumProductReferenceNode = xmlNode->FirstChildElement("premiumProductReference");

   if(premiumProductReferenceNode){premiumProductReferenceIsNull_ = false;}
   else{premiumProductReferenceIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- premiumProductReferenceNode , address : " << premiumProductReferenceNode << std::endl;
   #endif
   if(premiumProductReferenceNode)
   {
      if(premiumProductReferenceNode->Attribute("href") || premiumProductReferenceNode->Attribute("id"))
      {
          if(premiumProductReferenceNode->Attribute("id"))
          {
             premiumProductReferenceIDRef_ = premiumProductReferenceNode->Attribute("id");
             premiumProductReference_ = boost::shared_ptr<ProductReference>(new ProductReference(premiumProductReferenceNode));
             premiumProductReference_->setID(premiumProductReferenceIDRef_);
             IDManager::instance().setID(premiumProductReferenceIDRef_,premiumProductReference_);
          }
          else if(premiumProductReferenceNode->Attribute("href")) { premiumProductReferenceIDRef_ = premiumProductReferenceNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { premiumProductReference_ = boost::shared_ptr<ProductReference>(new ProductReference(premiumProductReferenceNode));}
   }

   //issuerNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* issuerNode = xmlNode->FirstChildElement("issuer");

   if(issuerNode){issuerIsNull_ = false;}
   else{issuerIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- issuerNode , address : " << issuerNode << std::endl;
   #endif
   if(issuerNode)
   {
      if(issuerNode->Attribute("href") || issuerNode->Attribute("id"))
      {
          if(issuerNode->Attribute("id"))
          {
             issuerIDRef_ = issuerNode->Attribute("id");
             issuer_ = boost::shared_ptr<IssuerId>(new IssuerId(issuerNode));
             issuer_->setID(issuerIDRef_);
             IDManager::instance().setID(issuerIDRef_,issuer_);
          }
          else if(issuerNode->Attribute("href")) { issuerIDRef_ = issuerNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { issuer_ = boost::shared_ptr<IssuerId>(new IssuerId(issuerNode));}
   }

   //tradeIdNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* tradeIdNode = xmlNode->FirstChildElement("tradeId");

   if(tradeIdNode){tradeIdIsNull_ = false;}
   else{tradeIdIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- tradeIdNode , address : " << tradeIdNode << std::endl;
   #endif
   if(tradeIdNode)
   {
      if(tradeIdNode->Attribute("href") || tradeIdNode->Attribute("id"))
      {
          if(tradeIdNode->Attribute("id"))
          {
             tradeIdIDRef_ = tradeIdNode->Attribute("id");
             tradeId_ = boost::shared_ptr<TradeId>(new TradeId(tradeIdNode));
             tradeId_->setID(tradeIdIDRef_);
             IDManager::instance().setID(tradeIdIDRef_,tradeId_);
          }
          else if(tradeIdNode->Attribute("href")) { tradeIdIDRef_ = tradeIdNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { tradeId_ = boost::shared_ptr<TradeId>(new TradeId(tradeIdNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<ProductReference> ProductComponentIdentifier::getPremiumProductReference()
{
   if(!this->premiumProductReferenceIsNull_){
        if(premiumProductReferenceIDRef_ != ""){
             return boost::shared_static_cast<ProductReference>(IDManager::instance().getID(premiumProductReferenceIDRef_));
        }else{
             return this->premiumProductReference_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<ProductReference>();
   }
}
boost::shared_ptr<IssuerId> ProductComponentIdentifier::getIssuer()
{
   if(!this->issuerIsNull_){
        if(issuerIDRef_ != ""){
             return boost::shared_static_cast<IssuerId>(IDManager::instance().getID(issuerIDRef_));
        }else{
             return this->issuer_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<IssuerId>();
   }
}
boost::shared_ptr<TradeId> ProductComponentIdentifier::getTradeId()
{
   if(!this->tradeIdIsNull_){
        if(tradeIdIDRef_ != ""){
             return boost::shared_static_cast<TradeId>(IDManager::instance().getID(tradeIdIDRef_));
        }else{
             return this->tradeId_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<TradeId>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-doc-5-4/ProductComponentIdentifier.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class ProductComponentIdentifier : ISerialized
    {
        public ProductComponentIdentifier(XmlNode xmlNode)
        {
            XmlNode premiumProductReferenceNode = xmlNode.SelectSingleNode("premiumProductReference");
            
            if (premiumProductReferenceNode != null)
            {
                if (premiumProductReferenceNode.Attributes["href"] != null || premiumProductReferenceNode.Attributes["id"] != null) 
                {
                    if (premiumProductReferenceNode.Attributes["id"] != null) 
                    {
                        premiumProductReferenceIDRef_ = premiumProductReferenceNode.Attributes["id"].Value;
                        ProductReference ob = new ProductReference(premiumProductReferenceNode);
                        IDManager.SetID(premiumProductReferenceIDRef_, ob);
                    }
                    else if (premiumProductReferenceNode.Attributes["href"] != null)
                    {
                        premiumProductReferenceIDRef_ = premiumProductReferenceNode.Attributes["href"].Value;
                    }
                    else
                    {
                        premiumProductReference_ = new ProductReference(premiumProductReferenceNode);
                    }
                }
                else
                {
                    premiumProductReference_ = new ProductReference(premiumProductReferenceNode);
                }
            }
            
        
            XmlNode issuerNode = xmlNode.SelectSingleNode("issuer");
            
            if (issuerNode != null)
            {
                if (issuerNode.Attributes["href"] != null || issuerNode.Attributes["id"] != null) 
                {
                    if (issuerNode.Attributes["id"] != null) 
                    {
                        issuerIDRef_ = issuerNode.Attributes["id"].Value;
                        IssuerId ob = new IssuerId(issuerNode);
                        IDManager.SetID(issuerIDRef_, ob);
                    }
                    else if (issuerNode.Attributes["href"] != null)
                    {
                        issuerIDRef_ = issuerNode.Attributes["href"].Value;
                    }
                    else
                    {
                        issuer_ = new IssuerId(issuerNode);
                    }
                }
                else
                {
                    issuer_ = new IssuerId(issuerNode);
                }
            }
            
        
            XmlNode tradeIdNode = xmlNode.SelectSingleNode("tradeId");
            
            if (tradeIdNode != null)
            {
                if (tradeIdNode.Attributes["href"] != null || tradeIdNode.Attributes["id"] != null) 
                {
                    if (tradeIdNode.Attributes["id"] != null) 
                    {
                        tradeIdIDRef_ = tradeIdNode.Attributes["id"].Value;
                        TradeId ob = new TradeId(tradeIdNode);
                        IDManager.SetID(tradeIdIDRef_, ob);
                    }
                    else if (tradeIdNode.Attributes["href"] != null)
                    {
                        tradeIdIDRef_ = tradeIdNode.Attributes["href"].Value;
                    }
                    else
                    {
                        tradeId_ = new TradeId(tradeIdNode);
                    }
                }
                else
                {
                    tradeId_ = new TradeId(tradeIdNode);
                }
            }
            
        
        }
        
    
        #region PremiumProductReference_
        private ProductReference premiumProductReference_;
        public ProductReference PremiumProductReference_
        {
            get
            {
                if (this.premiumProductReference_ != null)
                {
                    return this.premiumProductReference_; 
                }
                else if (this.premiumProductReferenceIDRef_ != null)
                {
                    premiumProductReference_ = IDManager.getID(premiumProductReferenceIDRef_) as ProductReference;
                    return this.premiumProductReference_; 
                }
                else
                {
                      return this.premiumProductReference_; 
                }
            }
            set
            {
                if (this.premiumProductReference_ != value)
                {
                    this.premiumProductReference_ = value;
                }
            }
        }
        #endregion
        
        public string premiumProductReferenceIDRef_ { get; set; }
        #region Issuer_
        private IssuerId issuer_;
        public IssuerId Issuer_
        {
            get
            {
                if (this.issuer_ != null)
                {
                    return this.issuer_; 
                }
                else if (this.issuerIDRef_ != null)
                {
                    issuer_ = IDManager.getID(issuerIDRef_) as IssuerId;
                    return this.issuer_; 
                }
                else
                {
                      return this.issuer_; 
                }
            }
            set
            {
                if (this.issuer_ != value)
                {
                    this.issuer_ = value;
                }
            }
        }
        #endregion
        
        public string issuerIDRef_ { get; set; }
        #region TradeId_
        private TradeId tradeId_;
        public TradeId TradeId_
        {
            get
            {
                if (this.tradeId_ != null)
                {
                    return this.tradeId_; 
                }
                else if (this.tradeIdIDRef_ != null)
                {
                    tradeId_ = IDManager.getID(tradeIdIDRef_) as TradeId;
                    return this.tradeId_; 
                }
                else
                {
                      return this.tradeId_; 
                }
            }
            set
            {
                if (this.tradeId_ != value)
                {
                    this.tradeId_ = value;
                }
            }
        }
        #endregion
        
        public string tradeIdIDRef_ { get; set; }
        
    
        
    
    }
    
}

